viadot/sources/supermetrics.py

```diff
@@ -0,0 +1,164 @@
+from ..utils import handle_api_response
+from .base import Source
+
+
+class Supermetrics(Source):
+    """
+    A class implementing the Supermetrics API.
+
+    Documentation for this API is located at: https://supermetrics.com/docs/product-api-getting-started/
+    Usage limits: https://supermetrics.com/docs/product-api-usage-limits/
+
+    Parameters
+    ----------
+    query_params : Dict[str, Any], optional
+        The parameters to pass to the GET query.
+        See https://supermetrics.com/docs/product-api-get-data/ for full specification,
+        by default None
+    """
+
+    API_ENDPOINT = "https://api.supermetrics.com/enterprise/v2/query/data/json"
+
+    def __init__(self, *args, query_params: Dict[str, Any] = None, **kwargs):
+        DEFAULT_CREDENTIALS = get_source_credentials("supermetrics")
+        credentials = kwargs.pop("credentials", DEFAULT_CREDENTIALS)
+
+        super().__init__(*args, credentials=credentials, **kwargs)
+
+        if not self.credentials:
+            self.logger.debug(
+                "Credentials not specified. Falling back to `boto3` default credentials."
+            )
+
+        self.query_params = query_params
+
+    @classmethod
+    def get_params_from_api_query(cls, url: str) -> Dict[str, Any]:
+        """Returns parmeters from API query in a dictionary"""
+        url_unquoted = urllib.parse.unquote(url)
+        s = urllib.parse.parse_qs(url_unquoted)
+        endpoint = list(s.keys())[0]
+        params = s[endpoint][0]
+        params_d = json.loads(params)
+        return params_d
+
+    @classmethod
+    def from_url(cls, url: str, credentials: Dict[str, Any] = None):
+        obj = Supermetrics(
+            credentials=credentials or get_source_credentials("supermetrics")
+        )
+        params = cls.get_params_from_api_query(url)
+        obj.query_params = params
+        return obj
+
+    def to_json(self, timeout=(3.05, 60 * 30)) -> Dict[str, Any]:
+        """Download query results to a dictionary.
+        Note that Supermetrics API will sometimes hang and not return any error message,
+        so we're adding a timeout to GET.
+
+        See [requests docs](https://docs.python-requests.org/en/master/user/advanced/#timeouts)
+        for an explanation of why this timeout value will work on long-running queries but fail fast
+        on connection issues.
+        """
+
+        if not self.query_params:
+            raise ValueError("Please build the query first")
+
+        params = {"json": json.dumps(self.query_params)}
+        headers = {"Authorization": f'Bearer {self.credentials.get("api_key")}'}
+
+        response = handle_api_response(
+            url=self.API_ENDPOINT, params=params, headers=headers, timeout=timeout
+        )
+        return response.json()
+
+    @classmethod
+    def _get_col_names_google_analytics(
+        cls,
+        response: dict,
+    ) -> List[str]:
+        """Returns list of Google Analytics columns names"""
```

Comment:
I think we can standardize docstrings as in other sources https://github.com/dyvenia/viadot/blob/9239fcbf185f93230407fd16b4e96cd081e3d09b/viadot/sources/redshift_spectrum.py#L109